olm/operator-bundle.md

```diff
@@ -0,0 +1,270 @@
+
+# output:
+#  - test/Dockerfile
+#  - test/annotations.yaml
+```
+
+Once the scaffolding is generated the user can do a `docker build` to create an operator bundle image.
+
+#### Validate an Operator Bundle
+As an operator author I want to validate an operator bundle image so that I can ensure it runs as expected on a cluster.
+```bash
+$ operator-framework bundle validate --image=quay.io/test/test-operator:v1
+```
+
+The validate command will do the following:
+* Make sure the image `label` and `annotations.yaml` are appropriately configured. If there is any mismatch, the tool shoud generate apropriate error message. 
+* Verify that the format of the bundle is valid. If the bundle is of `registry` format, we should verify that the bundle conforms to operator-registry standards.
+
+#### Run the Operator from the Bundle Image
+As an operator author I want to run my operator directly from the bundle image. Once an operator is packaged into a bundle image, we want to give the author ability to run it using `olm` directly from the bundle image.
+```bash
+# The following creates an 'Operator' CR managed by olm.
+cat <<EOF | kubectl apply -f -
+apiVersion: operators.operatorframework.io/v2alpha1
+kind: Operator
+metadata:
+  name: test-operator
+spec:
+  bundle:
+    image: quay.io/test/test-operator:v1
+EOF
+```
+
+Below is an example of how an operator bundle image can be unpacked to apply the manifests on a cluster.
+```bash
+$ docker save quay.io/test/test-operator:v1 -o bundle.tar
+$ tar -xvf bundle.tar
+
+$ tar -tf bundle.tar 
+39d24aee3ad2e8720c12042d5b9ba52ce14a12ed72815a759b41b01b9a8dbc03/
+39d24aee3ad2e8720c12042d5b9ba52ce14a12ed72815a759b41b01b9a8dbc03/VERSION
+39d24aee3ad2e8720c12042d5b9ba52ce14a12ed72815a759b41b01b9a8dbc03/json
+39d24aee3ad2e8720c12042d5b9ba52ce14a12ed72815a759b41b01b9a8dbc03/layer.tar
+58b4c261195b83bc0b12b80b63f8e11fb97b5d369aea80ca7cc558793bb507a0.json
+7b590145954570b3b3b52db41d4fa8950eefed80fd01c937fb3949b863fe0ede/
+7b590145954570b3b3b52db41d4fa8950eefed80fd01c937fb3949b863fe0ede/VERSION
+7b590145954570b3b3b52db41d4fa8950eefed80fd01c937fb3949b863fe0ede/json
+7b590145954570b3b3b52db41d4fa8950eefed80fd01c937fb3949b863fe0ede/layer.tar
+manifest.json
+repositories
+
+# The first image layer contains the manifest resources that
+# can be applied directly to a cluster.
+$ cat manifest.json | jq '.[0].Layers[0]'
+"7b590145954570b3b3b52db41d4fa8950eefed80fd01c937fb3949b863fe0ede/layer.tar"
+
+$ cat manifest.json | jq '.[0].Layers[0]' | xargs tar -xvf
+$ kubectl apply -f ./manifests
+```
+#### Run the Operator from the Bundle Folder
+This applies to an `operator-registry` bundle. As an operator author I want to apply a bundle folder directly on the cluster so that:
+* I can test my changes.
+* I can iterate faster.
+
+```bash
+tree test
+test
+├── 0.1.0
+│   ├── testbackup.crd.yaml
+│   ├── testcluster.crd.yaml
+│   ├── testoperator.v0.1.0.clusterserviceversion.yaml
```

Comment:
Do CSVs still include the "placeholder" namespace? This seems to assume that's not a problem in this example.

Reply:
We have a separate effort to enable "directly applying CSVs" that this section assumes is already in place.

But I think the example probably should include a namespace flag?